bennu-core/src/main/java/org/fenixedu/bennu/core/security/Authenticate.java

```diff
@@ -84,64 +97,80 @@ public static User login(HttpServletRequest request, HttpServletResponse respons
      * @param response
      *            The response associated with the requires that triggered the logout
      */
-    public static void logout(HttpServletRequest request, HttpServletResponse response) {
-        HttpSession session = request.getSession(false);
+    public static void logout(final HttpServletRequest request, final HttpServletResponse response) {
+        final HttpSession session = request.getSession(false);
         if (session != null) {
             User user = (User) session.getAttribute(LOGGED_USER_ATTRIBUTE);
             if (user != null) {
                 fireLogoutListeners(request, response, user);
             }
             session.invalidate();
         }
-        loggedUser.set(null);
+        clear();
     }
 
-    public static void mock(User user) {
-        loggedUser.set(user);
+    public static void mock(final User user, final String authenticationMethod) {
+        final AuthenticationContext authenticationContext = loggedUserContext.get();
+        if (authenticationContext == null || authenticationContext.getUser() != user) {
+            loggedUserContext.set(new AuthenticationContext(user, authenticationMethod));
+        } else {
+            authenticationContext.addAuthenticationMethodEvent(authenticationMethod);
+        }
     }
 
     public static void unmock() {
-        loggedUser.set(null);
+        clear();
     }
 
     public static User getUser() {
-        return loggedUser.get();
+        final AuthenticationContext context = loggedUserContext.get();
+        return context == null ? null : context.getUser();
     }
 
     public static boolean isLogged() {
-        return loggedUser.get() != null;
+        final AuthenticationContext context = loggedUserContext.get();
+        return context != null && context.getUser() != null;
     }
 
-    static void updateFromSession(HttpSession session) {
-        User user = (User) (session == null ? null : session.getAttribute(LOGGED_USER_ATTRIBUTE));
+    private static void storeInSession(final HttpServletRequest request, final AuthenticationContext authenticationContext) {
+        final HttpSession session = request.getSession();
+        session.setAttribute(LOGGED_USER_ATTRIBUTE, authenticationContext.getUser());
+        session.setAttribute(LOGGED_USER_AUTHENTICATION_CONTEXT_ATTRIBUTE, authenticationContext.export());
```

Comment:
why not store the authenticationContext in the session and avoid json des/serialisation ?

Reply:
The result of serializing the AuthenticationContext object would result in about 650 - 800 bytes.
Serializing the corresponding json String only occupies around 50 - 70 bytes.
Since the HttpSession may need to be passed between application servers (for example if the session is replicated in a cluster) then reducing the amount of data in the session may have a significant impact in overall application performance.